[8.19] [Theme] fix `theme$` usage when used with `useObservable` (#220141)

nickofthyme opened the pull request:
# Backport

This will backport the following commits from `main` to `8.19`:
 - https://github.com/elastic/kibana/pull/220141
 - https://github.com/elastic/kibana/pull/221979
 - https://github.com/elastic/kibana/pull/222099

nickofthyme committed:
[Theme] fix `theme$` usage when used with `useObservable` (#220141)

## Summary

When using `theme$` with
[`useObservable`](https://github.com/streamich/react-use/blob/ad33f76dfff7ddb041a9ef74b80656a94affaa80/src/useObservable.ts#L10-L21)
passing custom default `theme`, the resulting `theme` can be wrong.

Take for example this code...

```ts
import useObservable from 'react-use/lib/useObservable';

const theme = useObservable<CoreTheme>(kibanaTheme.theme$, {
  darkMode: false,
  name: 'amsterdam',
});
```

In such case `kibanaTheme.theme$` has the correct value but the
`useObservable` returns the default/initial value immediately, so the
default is always applied then updated, requiring 2 renderings just to
update to the correct theme.

The simplest approach to fix this is just to pass the
`kibanaTheme.getTheme()` as the default when using with `useObservable`.

```ts
const theme = useObservable<CoreTheme>(kibanaTheme.theme$, kibanaTheme.getTheme());
```

---

Ideally, in the future we have a commonly shared way to access the theme
in react via a `useKibanaTheme` hook or a better/more consistent API for
`useEuiTheme`.

### Checklist

- [x] [Unit or functional
tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html)
were updated or added to match the most common scenarios
- [x] The PR description includes the appropriate Release Notes section,
and the correct `release_note:*` label is applied per the
[guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>
Co-authored-by: Edgar Santos <edgar.santos@elastic.co>
(cherry picked from commit 12b7429afe0fb470fa6c85168c9040a5957bbe8b)

# Conflicts:
#	src/platform/plugins/private/kibana_overview/public/components/overview/overview.tsx
#	src/platform/plugins/shared/chart_expressions/expression_partition_vis/tsconfig.json
#	src/platform/plugins/shared/chart_expressions/expression_tagcloud/tsconfig.json
#	src/platform/plugins/shared/chart_expressions/expression_xy/tsconfig.json
#	src/platform/plugins/shared/guided_onboarding/public/components/guide_panel.tsx
#	src/platform/plugins/shared/kibana_react/public/dark_mode/use_dark_mode.test.tsx
#	x-pack/platform/plugins/shared/fleet/public/applications/fleet/app.tsx
#	x-pack/platform/plugins/shared/fleet/public/applications/integrations/app.tsx
#	x-pack/platform/plugins/shared/fleet/public/applications/integrations/sections/epm/components/installation_status.tsx
#	x-pack/platform/plugins/shared/lens/public/editor_frame_service/editor_frame/workspace_panel/workspace_panel.tsx
#	x-pack/platform/plugins/shared/lens/public/visualizations/datatable/components/table_basic.tsx
#	x-pack/platform/plugins/shared/lens/public/visualizations/datatable/visualization.tsx
#	x-pack/solutions/observability/plugins/apm/public/components/app/service_dashboards/empty_dashboards.tsx
#	x-pack/solutions/observability/plugins/apm/public/components/app/service_overview/service_overview_instances_table/intance_details.tsx
#	x-pack/solutions/observability/plugins/apm/public/components/shared/charts/timeseries_chart.tsx
#	x-pack/solutions/observability/plugins/apm/public/components/shared/service_icons/index.tsx
#	x-pack/solutions/observability/plugins/infra/public/hooks/use_is_dark_mode.ts
#	x-pack/solutions/security/plugins/security_solution/public/common/mock/endpoint/app_root_provider.tsx
#	x-pack/solutions/security/plugins/security_solution/public/onboarding/components/onboarding_header/onboarding_header.tsx
#	x-pack/solutions/security/plugins/security_solution/public/overview/components/host_overview/index.tsx
#	x-pack/solutions/security/plugins/security_solution/tsconfig.json

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/302866)
* Commit: 68075ddbc1e3eda650dd7b82dc8953a703d0d286

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/302866#019713a0-d48d-4bf4-bfe8-ed8deef0d512)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/302866","number":302866,"commit":"68075ddbc1e3eda650dd7b82dc8953a703d0d286"}],"number":302866}
buildkite-pr-comment-->

nickofthyme committed:
chore: fix old deleted hook usage

nickofthyme committed:
Merge branch '8.19' into backport/8.19/pr-220141

nickofthyme committed:
chore: fix more type errors

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/notice'

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/303231)
* Commit: 45ef5edbe001650e665f4bc1342a38977a7c7085

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/303231#019718ca-aa87-405d-b35a-8413fe8991c6)

### History
* :broken_heart: [Build #303226](https://buildkite.com/elastic/kibana-pull-request/builds/303226) failed 4d6b7890432e80c510eaed0b824f4a52641d829c
* :broken_heart: [Build #303210](https://buildkite.com/elastic/kibana-pull-request/builds/303210) failed 367a90288e35bc3a85858155faab35b39cb8ed2f
* :broken_heart: [Build #302866](https://buildkite.com/elastic/kibana-pull-request/builds/302866) failed 68075ddbc1e3eda650dd7b82dc8953a703d0d286

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303231","number":303231,"commit":"45ef5edbe001650e665f4bc1342a38977a7c7085"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303226","number":303226,"commit":"4d6b7890432e80c510eaed0b824f4a52641d829c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303210","number":303210,"commit":"367a90288e35bc3a85858155faab35b39cb8ed2f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/302866","number":302866,"commit":"68075ddbc1e3eda650dd7b82dc8953a703d0d286"}],"number":303231}
buildkite-pr-comment-->

nickofthyme committed:
[Integrations] Fix styled component theme lookup issue (#221979)

## Summary

This fixes an issue caused by changes in #220141, specifically [here](https://github.com/elastic/kibana/pull/220141/files#diff-90215c9231256571f4b7369e58505a234b7f7b4082cf7c9fda47692787b52222L80-L83), which prevented the style components from receiving the correct `colorMode`.

nickofthyme committed:
Merge branch '8.19' into backport/8.19/pr-220141

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/styled_components_mapping'

nickofthyme committed:
chore: revert deletion of `useDarkMode`

This is needed in cases where the color mode is being passed to the `ThemeProvider` for styled-components

nickofthyme committed:
chore: replace usages for `useDarkMode`

nickofthyme committed:
Merge branch '8.19' into backport/8.19/pr-220141

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/303816)
* Commit: 5479137481f83ba3fd2c2845130b3c0ee6689db6

### Failed CI Steps
* [Scout: [ platform / streams_app ] plugin](https://buildkite.com/elastic/kibana-pull-request/builds/303816#0197226f-7462-4c82-ae85-a5fe7c65ba39)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303816#0197226f-7462-4c82-ae85-a5fe7c65ba39) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303816/jobs/0197226f-7462-4c82-ae85-a5fe7c65ba39/artifacts/01972276-fb72-4020-9e67-a50a6a88251c) Scout: [ platform / streams_app ] plugin / stateful - Classic Streams - full flow
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303816#0197226f-7462-4c82-ae85-a5fe7c65ba39) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303816/jobs/0197226f-7462-4c82-ae85-a5fe7c65ba39/artifacts/01972276-fb72-4ddb-81ec-5d22bf4ba9c2) Scout: [ platform / streams_app ] plugin / stateful - Wired Streams - full flow

### History
* :broken_heart: [Build #303576](https://buildkite.com/elastic/kibana-pull-request/builds/303576) failed d35ad775b125cca47305723cfa4922a716a0b116
* :broken_heart: [Build #303231](https://buildkite.com/elastic/kibana-pull-request/builds/303231) failed 45ef5edbe001650e665f4bc1342a38977a7c7085
* :broken_heart: [Build #303226](https://buildkite.com/elastic/kibana-pull-request/builds/303226) failed 4d6b7890432e80c510eaed0b824f4a52641d829c
* :broken_heart: [Build #303210](https://buildkite.com/elastic/kibana-pull-request/builds/303210) failed 367a90288e35bc3a85858155faab35b39cb8ed2f
* :broken_heart: [Build #302866](https://buildkite.com/elastic/kibana-pull-request/builds/302866) failed 68075ddbc1e3eda650dd7b82dc8953a703d0d286

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303816","number":303816,"commit":"5479137481f83ba3fd2c2845130b3c0ee6689db6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303576","number":303576,"commit":"d35ad775b125cca47305723cfa4922a716a0b116"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303231","number":303231,"commit":"45ef5edbe001650e665f4bc1342a38977a7c7085"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303226","number":303226,"commit":"4d6b7890432e80c510eaed0b824f4a52641d829c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303210","number":303210,"commit":"367a90288e35bc3a85858155faab35b39cb8ed2f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/302866","number":302866,"commit":"68075ddbc1e3eda650dd7b82dc8953a703d0d286"}],"number":303816}
buildkite-pr-comment-->

nickofthyme committed:
Merge branch '8.19' into backport/8.19/pr-220141

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/303888)
* Commit: a670770740367bbcd84977edc7b7746ed1f7366c

### Failed CI Steps
* [FTR Configs #16](https://buildkite.com/elastic/kibana-pull-request/builds/303888#01972357-811a-4b84-aec8-60bed6da97f9)
* [FTR Configs #22](https://buildkite.com/elastic/kibana-pull-request/builds/303888#01972357-811e-4375-8f5f-f7aee39dfbb2)
* [FTR Configs #22](https://buildkite.com/elastic/kibana-pull-request/builds/303888#0197237c-75a0-4615-992b-ba3b2080a5cb)
* [Fleet Cypress Tests #1](https://buildkite.com/elastic/kibana-pull-request/builds/303888#0197233e-854e-4a41-998d-2a3a268221ae)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303888#01972357-811e-4375-8f5f-f7aee39dfbb2) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303888/jobs/01972357-811e-4375-8f5f-f7aee39dfbb2/artifacts/0197237c-4d7c-439a-9e17-2d3c377243e3) FTR Configs #<span></span>22 / Embeddable Alerts Table Embeddable alerts panel "before all" hook for "should only show alerts from the observability area (o11y+stack) when selecting it"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303888#0197237c-75a0-4615-992b-ba3b2080a5cb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303888/jobs/0197237c-75a0-4615-992b-ba3b2080a5cb/artifacts/01972386-4242-4a8d-bf88-01afc6c4ceda) FTR Configs #<span></span>22 / Embeddable Alerts Table Embeddable alerts panel "before all" hook for "should only show alerts from the observability area (o11y+stack) when selecting it"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303888#01972357-811a-4b84-aec8-60bed6da97f9) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303888/jobs/01972357-811a-4b84-aec8-60bed6da97f9/artifacts/01972379-1c2d-4270-9fc0-0ebc44bf3e52) FTR Configs #<span></span>16 / InfraOps App Metrics UI Node Details #<span></span>Asset Type: host Logs Tab preserves search term between page reloads
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303888#0197233e-854e-4a41-998d-2a3a268221ae) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303888/jobs/0197233e-854e-4a41-998d-2a3a268221ae/artifacts/0197237a-7d66-4121-b5e4-5912c004cba1) Fleet Cypress Tests #<span></span>1 / View agents list Agent status filter should filter on healthy (16 result)
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303888#0197233e-854e-4a41-998d-2a3a268221ae) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303888/jobs/0197233e-854e-4a41-998d-2a3a268221ae/artifacts/0197237a-7d66-48e5-9772-78f3f6aaf9c0) Fleet Cypress Tests #<span></span>1 / View agents list Bulk actions should allow to bulk upgrade agents and cancel that upgrade

### History
* :broken_heart: [Build #303816](https://buildkite.com/elastic/kibana-pull-request/builds/303816) failed 5479137481f83ba3fd2c2845130b3c0ee6689db6
* :broken_heart: [Build #303576](https://buildkite.com/elastic/kibana-pull-request/builds/303576) failed d35ad775b125cca47305723cfa4922a716a0b116
* :broken_heart: [Build #303231](https://buildkite.com/elastic/kibana-pull-request/builds/303231) failed 45ef5edbe001650e665f4bc1342a38977a7c7085
* :broken_heart: [Build #303226](https://buildkite.com/elastic/kibana-pull-request/builds/303226) failed 4d6b7890432e80c510eaed0b824f4a52641d829c
* :broken_heart: [Build #303210](https://buildkite.com/elastic/kibana-pull-request/builds/303210) failed 367a90288e35bc3a85858155faab35b39cb8ed2f

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303888","number":303888,"commit":"a670770740367bbcd84977edc7b7746ed1f7366c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303816","number":303816,"commit":"5479137481f83ba3fd2c2845130b3c0ee6689db6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303576","number":303576,"commit":"d35ad775b125cca47305723cfa4922a716a0b116"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303231","number":303231,"commit":"45ef5edbe001650e665f4bc1342a38977a7c7085"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303226","number":303226,"commit":"4d6b7890432e80c510eaed0b824f4a52641d829c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303210","number":303210,"commit":"367a90288e35bc3a85858155faab35b39cb8ed2f"}],"number":303888}
buildkite-pr-comment-->

nickofthyme committed:
Merge branch '8.19' into backport/8.19/pr-220141

ppisljar reviewed (approved):
LGTM
